### ISSUE TITLE
Crash when running a query

### ISSUE THREAD

**Issue by lucatorella** [couchbase/couchbase-lite-ios#1282]
In my crash reports I saw a crash coming possibly from ForestDB. My find method, creates a query, set some settings, and then run the query like this `let rows = try query.run()`.

I don't think it matters but all this code is run from a `QOS_CLASS_USER_INITIATED` queue.

I'm not able to replicate it, so I can't give you more information.

```
Crashed: com.apple.root.user-initiated-qos
0  MyApp                          0x10055102c fdb_get_kvs_info + 25848
1  MyApp                          0x10051cf9c cbforest::KeyStore::name() const + 91664
2  MyApp                          0x10051bccc cbforest::Database::contains(cbforest::KeyStore&) const + 86848
3  MyApp                          0x100523004 cbforest::MapReduceIndex::updateDocInIndex(cbforest::Transaction&, cbforest::Mappable const&) + 116344
4  MyApp                          0x10052396c cbforest::MapReduceIndexer::addMappable(cbforest::Mappable const&) + 118752
5  MyApp                          0x100517a38 CocoaIndexer::addDocument(cbforest::Document const&) + 69804
6  MyApp                          0x1005236b8 cbforest::MapReduceIndexer::run() + 118060
7  MyApp                          0x100513f6c __41-[CBL_ForestDBViewStorage updateIndexes:]_block_invoke + 54752
8  MyApp                          0x100519eb4 couchbase_lite::tryStatus(CBLStatus () block_pointer) + 79144
9  MyApp                          0x100513b58 -[CBL_ForestDBViewStorage updateIndexes:] + 53708
10 MyApp                          0x10081c19c CBLStatusFromBulkDocsResponseItem + 2954856
11 MyApp                          0x10081c090 CBLStatusFromBulkDocsResponseItem + 2954588
12 MyApp                          0x1008438a4 CBLGetReduceFunc + 3116400
13 MyApp                          0x100841f98 CBLGetReduceFunc + 3109988
14 MyApp                          0x100152de4 CouchbaseJsApi.(find(String, conditions : [String : AnyObject], limit : UInt?, skip : UInt, completion : ([NSDictionary], JsError?) -> ()) -> ()).(closure #1) (Couchbase.swift:438)
15 MyApp      0x10014d544 partial apply for CouchbaseJsApi.(find(String, conditions : [String : AnyObject], limit : UInt?, skip : UInt, completion : ([NSDictionary], JsError?) -> ()) -> ()).(closure #1) (Couchbase.swift)
16 libdispatch.dylib              0x182fed4bc _dispatch_call_block_and_release + 24
17 libdispatch.dylib              0x182fed47c _dispatch_client_callout + 16
18 libdispatch.dylib              0x182ffb914 _dispatch_root_queue_drain + 2140
19 libdispatch.dylib              0x182ffb0b0 _dispatch_worker_thread3 + 112
20 libsystem_pthread.dylib        0x183205470 _pthread_wqthread + 1092
21 libsystem_pthread.dylib        0x183205020 start_wqthread + 4
```

Line 438 is the `let rows = try query.run()`

--------
* Version: 1.2.2
* Client OS: iOS 9.3.2
* Server: couchdb 1.6.1

**Comment by lucatorella** [comment]
Just got another that is quite similar

```
Crashed: com.apple.root.background-qos
0  MyApp                          0x1004daa90 _fdb_abort_transaction + 180484
1  MyApp                          0x1004c3ef8 cbforest::Database::endTransaction(cbforest::Transaction*) + 87404
2  MyApp                          0x1004c3d9c cbforest::Transaction::~Transaction() + 87056
3  MyApp                          0x1004cb764 cbforest::MapReduceIndexer::~MapReduceIndexer() + 118232
4  MyApp                          0x1004bc0f4 __41-[CBL_ForestDBViewStorage updateIndexes:]_block_invoke + 55144
5  MyApp                          0x1004c1eb4 couchbase_lite::tryStatus(CBLStatus () block_pointer) + 79144
6  MyApp                          0x1004bbb58 -[CBL_ForestDBViewStorage updateIndexes:] + 53708
7  MyApp                          0x1007c419c CBLStatusFromBulkDocsResponseItem + 2954856
8  MyApp                          0x1007c4090 CBLStatusFromBulkDocsResponseItem + 2954588
9  MyApp                          0x1007eb8a4 CBLGetReduceFunc + 3116400
10 MyApp                          0x1007e9f98 CBLGetReduceFunc + 3109988
11 MyApp                          0x1000f9700 CouchbaseJsApi.(updateViews() -> ()).(closure #1) (Couchbase.swift:206)
12 libdispatch.dylib              0x1834f94bc _dispatch_call_block_and_release + 24
13 libdispatch.dylib              0x1834f947c _dispatch_client_callout + 16
14 libdispatch.dylib              0x183507914 _dispatch_root_queue_drain + 2140
15 libdispatch.dylib              0x1835070b0 _dispatch_worker_thread3 + 112
16 libsystem_pthread.dylib        0x183711470 _pthread_wqthread + 1092
17 libsystem_pthread.dylib        0x183711020 start_wqthread + 4
```

again line 206, is another `try query.run()`

**Comment by snej** [comment]
Just to make sure: are you following the threading guidelines, i.e. using CBL objects only on the thread/queue on which they were created? Otherwise you're going to get random crashes like these.

(Also, your queue is serial, right?)

**Comment by lucatorella** [comment]
thank you I'll double check it

**Comment by lucatorella** [comment]
I read the documentation again and I checked our code and I have a question. 
In the "Concurrency support" section of the documentation, it says that one option is to use some special methods with background blocks, and another option is to use a separate thread (created via NSThread).
Do you really need a separate thread? Isn't enough to create a custom serial queue and use that queue when starting a query, starting the replication, etc.?

**Comment by lucatorella** [comment]
Let me rephrase. In the documentation you suggest to use something like this (by the way the Swift example contains some mistakes):

```
class DatabaseManager {
    ...
    private func createCouchbaseManager() {
        NSThread.detachNewThreadSelector(#selector(DatabaseManager.runBackground), toTarget: self, withObject: nil)
    }

    @objc private func runBackground() {
        let manager = CBLManager()
        let db = try! manager.databaseNamed("couchbase")
    }
    ...
}
```

But it should be perfectly fine to create the db and manager like this:
```
class DatabaseManager {
    private let couchbaseQueue = dispatch_queue_create("com.mydom.myapp", DISPATCH_QUEUE_SERIAL);

    private func createCouchbaseManager() {
        dispatch_async(couchbaseQueue) {
            let manager = CBLManager()
            let db = try! manager.databaseNamed("couchbase")
        }
    }
...
```
and then of course use always the same serial queue when doing queries.

I'm asking because queue != thread, and in the examples in the documentation you are using threads while usually in Cocoa it's always about GCD queues.

**Comment by pasin** [comment]
@lucatorella You could either use thread or queue. If you use a dispatch queue, you will need to set Manager.dispatchQueue property as well. I'm closing the issue now. Please let us know if after setting the dispatchQueue property, you are still encountering the issue or not. We are reopen the issue any time. Thanks.

**Comment by lucatorella** [comment]
oh I see. I as wrapping all couchbase code inside my `couchbaseQueue`, but I never set the `CBLManager` `dispatchQueue` property. I hope it will fix it. Thanks! Btw I think it would be great to update the documentation in the couchbase website, is much more useful to provide an example with the queue than with a `NSThread`

**Comment by eraser2021999** [comment]
Thanks for this thread, it really helped me. I had no idea about `dispatchQueue` property. Please try to update the documentation if it's possible.
